### PR TITLE
[GraphBolt][CUDA] Dataloader feature overlap fix

### LINKED ISSUE #7028: [GraphBolt][CUDA] GraphBolt MultiGPU accuracy with feature overlap.
## 🔨Work Item

**IMPORTANT:**
* This template is only for dev team to track project progress. For feature request or bug report, please use the corresponding issue templates.
* DO NOT create a new work item if the purpose is to fix an existing issue or feature request. We will directly use the issue in the project tracker.

Project tracker: https://github.com/orgs/dmlc/projects/2

## Description

Enabling the feature overlap optimization in the dataloader for the multi-GPU example degrades the accuracy. There seems to be a bug. We disabled this optimization for the multiGPU example as a workaround but the issue needs to be identified and resolved. Then we can reenable the overlap optimization.

@TristonC

### PR DESCRIPTION
**Pull request by mfbalin**
## Description
See the picture in #6983. The Awaiter block is missing. This is due to a bug in the wrap parent usage. Fixes #7028 and #7008. This bug arised due to a buggy use of the _find_and_wrap_parent function. It needs to return the modified datapipe_graph to the user so that the user uses the updated version after modifications.

Below, you can see that both Bufferer and Awaiter is present after modification. Awaiter basically synchronizes the feature copy on the uva_stream with the main stream. When it is missing, the fetched features will be accessed when the copy operation has not finished yet.

image

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by TristonC** [comment]
LGTM. And confirmed the accuracy is back to 0.7141 with an 8GPU test. And with the overlap feature, the performance improved about 17% for per epoch training time.

File: examples/multigpu/graphbolt/node_classification.py [hunk 1]
```diff
@@ -139,10 +139,7 @@ def create_dataloader(
     if args.storage_device == "cpu":
         datapipe = datapipe.copy_to(device)
 
-    # Until https://github.com/dmlc/dgl/issues/7008, overlap should be False.
-    dataloader = gb.DataLoader(
-        datapipe, args.num_workers, overlap_feature_fetch=False
-    )
+    dataloader = gb.DataLoader(datapipe, args.num_workers)
```
[the remark this answers is not in the document]
Comment (mfbalin) [reply examples/multigpu/graphbolt/node_classification.py:142]: It is already True by default, do you think it is still better to explicitly pass it as True?

[the remark this answers is not in the document]
Comment (TristonC) [reply examples/multigpu/graphbolt/node_classification.py:142]: Might not be a good idea to let user worry about the overlap_feature_fetch parameter. Need a unit test to avoid the accuracy issue come back again.

[the remark this answers is not in the document]
Comment (mfbalin) [reply examples/multigpu/graphbolt/node_classification.py:142]: Somehow, the regression test that we already have didn't catch this issue. The issue was reproducible only on machines where the feature fetch was the bottleneck. I will think of a way to write correctness unit tests.

[the remark this answers is not in the document]
Comment (Rhett-Ying) [reply examples/multigpu/graphbolt/node_classification.py:142]: could you elaborate the root cause in the PR description?

[the remark this answers is not in the document]
Comment (mfbalin) [reply examples/multigpu/graphbolt/node_classification.py:142]: It is already in the PR description.

[the remark this answers is not in the document]
Comment (mfbalin) [reply examples/multigpu/graphbolt/node_classification.py:142]: Further elaborating the main issue.

[the remark this answers is not in the document]
Comment (mfbalin) [reply examples/multigpu/graphbolt/node_classification.py:142]: I just noticed, I think the root cause is the buggy _find_and_wrap_parent implementation. It needs to modify datapipe_graph on return.

[the remark this answers is not in the document]
Comment (mfbalin) [reply examples/multigpu/graphbolt/node_classification.py:142]: Let me fix it in a better way instead of this hacky way.

**Comment by mfbalin** [comment]
@Rhett-Ying fixed the root cause, it was the buggy use of `_find_and_wrap_parent`. Because it modifies the datapipe_graph parameter it receives, it needs to return the modified version back to the user.

File: examples/multigpu/graphbolt/node_classification.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mfbalin) [reply examples/multigpu/graphbolt/node_classification.py:142]: Fixed.

File: python/dgl/graphbolt/dataloader.py [hunk 2]
```diff
@@ -221,25 +219,23 @@ def __init__(
...
-                Bufferer,
-                buffer_size=2,
-            )
-            _find_and_wrap_parent(
-                datapipe_graph,
-                datapipe_adjlist,
-                EndMarker,
-                Awaiter,
-            )
+            if len(feature_fetchers) > 0:
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/dataloader.py:222]: pls add dedicated testcases...

File: python/dgl/graphbolt/dataloader.py [hunk 3]
```diff
@@ -221,25 +219,23 @@ def __init__(
...
+                datapipe_graph = _find_and_wrap_parent(
+                    datapipe_graph,
+                    EndMarker,
+                    Bufferer,
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/dataloader.py:226]: and why not improve `_find_wrap_parent()` to support appending multiple datapipes in one call?

Comment (mfbalin) [reply python/dgl/graphbolt/dataloader.py:226]: How are we going to pass the arguments to `_find_wrap_parent()` for the different datapipes in a single call?

File: python/dgl/graphbolt/dataloader.py (hunk 2 above)
Comment (mfbalin) [reply python/dgl/graphbolt/dataloader.py:222]: Adding tests.

Comment (mfbalin) [reply python/dgl/graphbolt/dataloader.py:222]: Added.

File: python/dgl/graphbolt/dataloader.py (hunk 3 above)
Comment (mfbalin) [reply python/dgl/graphbolt/dataloader.py]: We are not using `_find_and_wrap_parent` anymore.

**Comment by mfbalin** [comment]
@Rhett-Ying I think I finally did it the right way. Should be ready to merge if it looks good to you.

File: tests/python/pytorch/graphbolt/test_dataloader.py [hunk 4]
```diff
@@ -70,13 +73,22 @@ def test_gpu_sampling_DataLoader(overlap_feature_fetch):
...
+        datapipe = dgl.graphbolt.FeatureFetcher(
+            datapipe,
+            feature_store,
+            ["a", "b"],
+        )
 
     dataloader = dgl.graphbolt.DataLoader(
         datapipe, overlap_feature_fetch=overlap_feature_fetch
     )
+    if enable_feature_fetch and overlap_feature_fetch:
```
Comment (Rhett-Ying) [inline tests/python/pytorch/graphbolt/test_dataloader.py:86]: it's better to check if `len(Awaiter)` and `len(Buffer)` equal with expected value: 0 or 1 for all combination of `overlap_xxx` and `enable_xxx`.

**Review by Rhett-Ying — APPROVED** [review]

File: tests/python/pytorch/graphbolt/test_dataloader.py (hunk 4 above)
Comment (mfbalin) [reply tests/python/pytorch/graphbolt/test_dataloader.py:86]: Got it.

**Review by frozenbugs — APPROVED** [review]

**Comment by frozenbugs** [comment]
For the discussion about whether we need overlap_feature_fetch, let's keep for a while after release, and clean it up if it is proven to be a stable solution for different hardware config.
